Improve `PortableHost{Collection,Object}` dictionary declarations

makortel opened the pull request:
#### PR description:

Testing ROOT PR on an option to disable header parsing during `TClass::GetClass()` call (https://github.com/root-project/root/pull/18402) it was noticed the type alias gets listed in the `rootmap` file only if the alias is requested before the real type (see https://github.com/root-project/root/issues/19705 for more details).

In the mean time, having the type alias in the `rootmap` file is necessary to avoid header parsing for the execution of the read rules that use the type alias names, e.g.
https://github.com/cms-sw/cmssw/blob/2de49449715cd97f6de642a9b73a06318bd3f8b9/DataFormats/Portable/interface/PortableHostCollectionReadRules.h#L84
, and therefore it seemed worthwhile to change the dictionaries with ` PortableHost{Collection,Object}`.

In the present state this PR demonstrates what the impact would be for the `PortableTestObjects`. If deemed viable, the next steps would be to update `DataFormats/Portable` README and `scripts`, and then update all the other `classes_def.xml` files that declare these portable data products.

Resolves https://github.com/cms-sw/framework-team/issues/1542

#### PR validation:

The `testHeterogeneousCoreAlpakaTestWriteReadSerialSync` unit test succeeds with the build of https://github.com/cms-sw/root/pull/222#issuecomment-3235701670. The resulting `rootmap` contains the `::Product`, `::Layout`, and `::Implementation` type aliases that the read rules use.

makortel committed:
List PortableTestHost{Collection,Object} nested type aliases before their aliased types in classes_def.xml

In ROOT presently the type alias gets listed in the rootmap file only
if the alias is requested before the real type. Having the type alias
in the rootmap file is necessary to avoid header parsing for the
execution of the read rules that use the type alias names.

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"cb4f96545ed26fa5d4db5611491c8d18f5ac584c":{"files":["DataFormats/PortableTestObjects/src/classes_def.xml"],"squashed":false,"time":1756492494}},"emoji":{},"last_seen_sha":"cb4f96545ed26fa5d4db5611491c8d18f5ac584c","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48824/45942

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-556bb6/47912/summary.html
**COMMIT**: cb4f96545ed26fa5d4db5611491c8d18f5ac584c
**CMSSW**: CMSSW_15_1_X_2025-08-29-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48824/47912/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-29-1100+556bb6/70052/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4115479
 - DQMHistoTests: [Total failures: 29](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-29-1100+556bb6/70052/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4115430
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-29-1100+556bb6/70052/triggerResults): no differences found

makortel commented:
@fwyzard Given 
* https://github.com/cms-sw/cmssw/issues/49333, fixed in https://github.com/cms-sw/cmssw/pull/49360
* https://github.com/cms-sw/cmssw/issues/49399, fixed in https://github.com/cms-sw/cmssw/pull/49422
* https://github.com/cms-sw/cmssw/issues/49470, fixed in https://github.com/cms-sw/cmssw/pull/49473

regardless of https://github.com/cms-sw/cmssw/issues/49458 we should probably revive this PR and go systematically through all dictionary definitions of `Portable*` types (and update documentation).

I have a vague recollection though there was some detail I was not happy with, but I can't remember what it was.

fwyzard commented:
For example, something I noticed earlier today is that for the (deprecated) cases where a class inherits from a PortableCollection, the read rule refers to the portable collection, not to the derived class.
I don't know what is the correct behaviour, but I found the current one confusing...

makortel committed:
Include the necessity of dictionaries for the nested type alias in README

makortel committed:
Define dictionaries of nested type aliases of Portable* before their aliased-to types

This should avoid ROOT header parsing

makortel committed:
Mark LST collections transient

They are not in DataFormats, so they are allowed to be transient only.

makortel committed:
Add class version and checksum for TestStruct

makortel commented on DataFormats/BeamSpot/src/classes_def.xml:
```diff
@@ -5,11 +5,17 @@
+
   <class name="BeamSpotPOD" ClassVersion="3">
    <version ClassVersion="3" checksum="280341519"/>
   </class>
```
@pcanal I think we discussed about the apparent need to define the class versions and checksums twice (the `BeamSpotHost::Product` is an alias for `BeamSpotPOD`). Or, I didn't manage to come up with any other way that would result in both type names to end up in the `.rootmap` file. I wouldn't hold this PR for an improvement, but we should follow up in https://github.com/cms-sw/cmssw/issues/49458.

makortel commented on DataFormats/ParticleFlowReco/src/classes_serial_def.xml:
```diff
@@ -1,29 +1,29 @@
   <class name="reco::CaloRecHitSoA::View"/>
-  <class name="reco::CaloRecHitHostCollection" ClassVersion="3">
-    <version ClassVersion="3" checksum="1876594952"/>
-  </class>
```
I also removed the explicit versions and checksums for the `PortableHostCollection`s that still had it. Their versioning should be done in a different way because `PortableHostCollection<T>` is a class template.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-556bb6/50438/summary.html
**COMMIT**: d001897e21f624c9387b0f5160264a236b16643d
**CMSSW**: CMSSW_16_1_X_2026-01-05-1100/el8_amd64_gcc13
**Additional Tests**: GPU,AMD_MI300X,AMD_W7900,NVIDIA_H100,NVIDIA_L40S,NVIDIA_T4
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/48824/50438/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_1_X_2026-01-05-1100+556bb6/72570/validateJR/logRootQA-events.log) to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_1_X_2026-01-05-1100+556bb6/72570/validateJR.html)
 - Reco comparison had  4 failed jobs
 - DQMHistoTests: Total files compared: 55
 - DQMHistoTests: Total histograms compared: 4513958
 - DQMHistoTests: [Total failures: 61](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_1_X_2026-01-05-1100+556bb6/72570/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4513877
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 54 files compared)
 - Checked 235 log files, 208 edm output root files, 55 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_1_X_2026-01-05-1100+556bb6/72570/triggerResults): no differences found


## AMD_W7900 Comparison Summary



**Summary**:
 - You potentially [added 24 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsAMD_W7900/CMSSW_16_1_X_2026-01-05-1100+556bb6/72571/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 251 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsAMD_W7900/CMSSW_16_1_X_2026-01-05-1100+556bb6/72571/validateJR.html)
 - Reco comparison had  6 failed jobs
 - DQMHistoTests: Total files compared: 11
 - DQMHistoTests: Total histograms compared: 149371
 - DQMHistoTests: [Total failures: 31271](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsAMD_W7900/CMSSW_16_1_X_2026-01-05-1100+556bb6/72571/validateJR.html)
 - DQMHistoTests: Total nulls: 9
 - DQMHistoTests: Total successes: 118091
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 10 files compared)
 - Checked 42 log files, 45 edm output root files, 11 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsAMD_W7900/CMSSW_16_1_X_2026-01-05-1100+556bb6/72571/triggerResults): no differences found


## NVIDIA_H100 Comparison Summary



**Summary**:
 - You potentially [removed 15 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_H100/CMSSW_16_1_X_2026-01-05-1100+556bb6/72573/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 258 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_H100/CMSSW_16_1_X_2026-01-05-1100+556bb6/72573/validateJR.html)
 - Reco comparison had  6 failed jobs
 - DQMHistoTests: Total files compared: 11
 - DQMHistoTests: Total histograms compared: 149371
 - DQMHistoTests: [Total failures: 28936](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_H100/CMSSW_16_1_X_2026-01-05-1100+556bb6/72573/validateJR.html)
 - DQMHistoTests: Total nulls: 7
 - DQMHistoTests: Total successes: 120428
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 10 files compared)
 - Checked 42 log files, 45 edm output root files, 11 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_H100/CMSSW_16_1_X_2026-01-05-1100+556bb6/72573/triggerResults): no differences found


## NVIDIA_L40S Comparison Summary



**Summary**:
 - You potentially [removed 4 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_L40S/CMSSW_16_1_X_2026-01-05-1100+556bb6/72574/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 253 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_L40S/CMSSW_16_1_X_2026-01-05-1100+556bb6/72574/validateJR.html)
 - Reco comparison had  6 failed jobs
 - DQMHistoTests: Total files compared: 11
 - DQMHistoTests: Total histograms compared: 149371
 - DQMHistoTests: [Total failures: 26087](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_L40S/CMSSW_16_1_X_2026-01-05-1100+556bb6/72574/validateJR.html)
 - DQMHistoTests: Total nulls: 9
 - DQMHistoTests: Total successes: 123275
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 10 files compared)
 - Checked 42 log files, 45 edm output root files, 11 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_L40S/CMSSW_16_1_X_2026-01-05-1100+556bb6/72574/triggerResults): no differences found


## NVIDIA_T4 Comparison Summary



**Summary**:
 - You potentially [removed 7 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_T4/CMSSW_16_1_X_2026-01-05-1100+556bb6/72572/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 256 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_T4/CMSSW_16_1_X_2026-01-05-1100+556bb6/72572/validateJR.html)
 - Reco comparison had  6 failed jobs
 - DQMHistoTests: Total files compared: 11
 - DQMHistoTests: Total histograms compared: 149371
 - DQMHistoTests: [Total failures: 27452](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_T4/CMSSW_16_1_X_2026-01-05-1100+556bb6/72572/validateJR.html)
 - DQMHistoTests: Total nulls: 11
 - DQMHistoTests: Total successes: 121908
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 10 files compared)
 - Checked 42 log files, 45 edm output root files, 11 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNVIDIA_T4/CMSSW_16_1_X_2026-01-05-1100+556bb6/72572/triggerResults): no differences found

fwyzard commented:
> I think it would be safest to merge it soon and backport to 16_0_X.

Agreed.


> Is #49734 planned to be backported to 16_0_X?

Yes, that would be my preference (on top of backport this PR).

fwyzard commented:
@cms-sw/reconstruction-l2 could you check and sign this PR 🙏🏻 ?
We should try to get it in CMSSW_16_1_0_pre1 this week 😎

fwyzard commented:
@cms-sw/simulation-l2 could you check and sign this PR 🙏🏻 ?
We should try to get it in CMSSW_16_1_0_pre1 this week 😎

civanch commented:
+1

jfernan2 commented:
+1

makortel commented:
> > Is #49734 planned to be backported to 16_0_X?
> 
> Yes, that would be my preference (on top of backport this PR).

Thanks for the clarification. Given that order of preference, I'll open a backport of this PR later today.

makortel commented:
> I'll open a backport of this PR later today.

Here https://github.com/cms-sw/cmssw/pull/49774 (I was able to use the same topic branch)

civanch commented:
@Moanwar, can you please, sign this old PR?

Moanwar commented:
+Upgrade
